### PR TITLE
[Feature] Support DyHead

### PR DESCRIPTION
**Pull request by shinya7y**
## Motivation

This PR aims to support DyHead (CVPR 2021).
Dynamic Head: Unifying Object Detection Heads with Attentions https://arxiv.org/abs/2106.08322
This will close https://github.com/open-mmlab/mmdetection/issues/6634

I trained ATSS+DyHead (R-50) only. Please train other models if needed.

Figure for README.md:
DyHead_fig1

## Checklist

- [x] Add DYReLU
- [x] Add DyHead
- [x] Add docstring
- [x] Add unit tests
- [x] Add method's README.md
- [x] Add metafile.yml and update model-index.yml
- [x] Update README.md, README_zh-CN.md
- [x] Align inference accuracy of ATSS+DyHead (R-50)
     [converted weight](https://github.com/shinya7y/weights/releases/download/v1.0.0/dyhead_r50_atss_fpn_1x_mmdet-4d5fec82.pth)
     box AP: 42.3 at 85ed81f (I think this is reasonable alignment without the need for changing existing code much.)
- [x] Align training accuracy of ATSS+DyHead (R-50)

| Method      | Backbone | code | config            | lr   | batch size | conv_offset | box AP |
|-------------|----------|-----------------|-|------|------------|--------|--------|
| ATSS+DyHead | R-50     | mmdet (8011283) | simple | 0.01 | 16 (4x4)   | zero init | 43.3       |
| ATSS+DyHead | R-50     | mmdet (c72c660) | repro | 0.01 | 16 (4x4)   | PyTorch default | 42.5       |
| ATSS+DyHead | R-50     | Detectron2      | - | 0.01 | 16 (8?x2?) | PyTorch default | 42.4   |
| ATSS+DyHead | R-50     | mrbench (paper) | - | 0.02 | ?  (8x?)   | ? | 42.6   |

### CODE REVIEW

**Comment by jshilong** [comment]
Thanks for your amazing work.
Does the table mean this version is higher than the original implementation?

**Comment by shinya7y** [comment]
Yes, the version (8011283) achieves higher AP.
Although the config `configs/dyhead/atss_r50_fpn_dyhead_1x_coco.py` will work as a minimum example to use DyHead in mmdet, it has many implementation differences. I will add another config to align accuracy.

File: mmdet/models/necks/dyhead.py
Comment (ZwwWayne) [inline mmdet/models/necks/dyhead.py:76]: I think we might not need DyDCNv2.
We can directly use ConvModule with `conv_cfg=dict(type='DCNv2')`

File: mmdet/models/utils/se_layer.py
Comment (ZwwWayne) [inline mmdet/models/utils/se_layer.py:62]: Shall we keep using DyReLU? It is consistent with DyHead and DyDCNv2

Comment (shinya7y) [reply mmdet/models/utils/se_layer.py:62]: I will change it. The original abbreviation is [DY-ReLU](https://arxiv.org/abs/2003.10027), but the authors also use [DyReLU](https://openaccess.thecvf.com/content/ICCV2021/papers/Dai_Dynamic_DETR_End-to-End_Object_Detection_With_Dynamic_Attention_ICCV_2021_paper.pdf).

File: configs/dyhead/README.md
Comment (ZwwWayne) [inline configs/dyhead/README.md:5]: The format of readme is updated. See changes in https://github.com/open-mmlab/mmdetection/pull/7051. Could you apply the newest change to this file?

File: configs/dyhead/README.md
Comment (ZwwWayne) [inline configs/dyhead/README.md:33]: Usually, we can keep citing the CVPR version in such a case. The paper link can be linked to the arxiv version so that readers can find the newest paper.

File: configs/dyhead/README.md
Comment (shinya7y) [reply configs/dyhead/README.md:5]: pre-commit does not pass before merging 7051.

**Review by RangiLyu — APPROVED** [review]
LGTM. By the way, should the models be uploaded to the openmmlab server before merging? @ZwwWayne

**Comment by hhaAndroid** [comment]
@shinya7y Please resolve the conflict.

**Comment by hhaAndroid** [comment]
@shinya7y Please resolve the conflict.

**Comment by ZwwWayne** [comment]
> LGTM. By the way, should the models be uploaded to the openmmlab server before merging? @ZwwWayne

Yes, we can do that. Is there any link so that I can download them and upload them to our model zoo server?

File: configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py
Comment (ZwwWayne) [inline configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py:1]: The name for_reproduction might be confusing. Can we keep the naming convention and rename it as atss_r50_caffe_fpn_dyhead_1x_coco.py?

File: configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py
Comment (ZwwWayne) [inline configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py:32]: suggest leaving comments to indicate it is different from that of the pytorch backbone.

File: configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py
Comment (ZwwWayne) [inline configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py:52]: suggest leaving comments to indicate it is different from that using Pytorch backbone

Comment (ZwwWayne) [reply configs/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco.py:52]: We should also comment other places if they are different from the config of using PyTorch backbone,

**Comment by ZwwWayne** [comment]
Hi @shinya7y ,
Thanks for the great work. This PR can be merged after resolving the last several comments. Let's make it in this release~

**Comment by shinya7y** [comment]
> > LGTM. By the way, should the models be uploaded to the openmmlab server before merging? @ZwwWayne
>
> Yes, we can do that. Is there any link so that I can download them and upload them to our model zoo server?

As written in README.md, the model weights are
https://github.com/shinya7y/weights/releases/download/v1.0.0/atss_r50_fpn_dyhead_for_reproduction_4x4_1x_coco_20220107_213939-162888e6.pth
https://github.com/shinya7y/weights/releases/download/v1.0.0/atss_r50_fpn_dyhead_4x4_1x_coco_20211219_023314-eaa620c6.pth

I will update the links if you could let me know new URLs.

File: configs/dyhead/README.md
Comment (ZwwWayne) [inline configs/dyhead/README.md:19]: https://openmmlab-share.oss-cn-hangzhou.aliyuncs.com/mmdetection/v2.0/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco/atss_r50_fpn_dyhead_for_reproduction_4x4_1x_coco_20220107_213939-162888e6.pth

https://openmmlab-share.oss-cn-hangzhou.aliyuncs.com/mmdetection/v2.0/dyhead/atss_r50_fpn_dyhead_for_reproduction_1x_coco/atss_r50_fpn_dyhead_for_reproduction_4x4_1x_coco_20220107_213939.log.json

File: configs/dyhead/README.md
Comment (ZwwWayne) [inline configs/dyhead/README.md:20]: https://openmmlab-share.oss-cn-hangzhou.aliyuncs.com/mmdetection/v2.0/dyhead/atss_r50_fpn_dyhead_4x4_1x_coco/atss_r50_fpn_dyhead_4x4_1x_coco_20211219_023314.log.json
https://openmmlab-share.oss-cn-hangzhou.aliyuncs.com/mmdetection/v2.0/dyhead/atss_r50_fpn_dyhead_4x4_1x_coco/atss_r50_fpn_dyhead_4x4_1x_coco_20211219_023314-eaa620c6.pth

**Comment by ZwwWayne** [comment]
Models have been uploaded.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**
